core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2022,29 +2022,80 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    def arbiter(
+      promise: Promise[Cause[E], List[B]],
+      buffer: Ref[Array[B]],
+      todo: Ref[Int],
+      idx: Int
+    )(res: Exit[E, B]): ZIO[R, Nothing, Unit] =
+      res.foldM[R, Nothing, Unit](
+        e => promise.fail(e).unit,
```

Comment:
@LGLO 
This has a different, lossy behavior towards failures than `zipWith` before. I wonder if you could store failures in another collection and merge them into one Cause at the end when all interrupts have ran (discarding uninteresting `Cause.interrupt` failures without additional finalizer failures)
Note: you could reuse a single Array by storing `Either` in `Array` cells and in case of failure take only Lefts.

Reply:
@neko-kai 
Thank you very much for that comment!

Now I'm wondering if previous behavior was intentional or just accidental because of implementation with `zipWithPar`. Effects run in concurrently so they can fail before receiving interruption (or be uninterruptible and complete) - simply we can have multiple failures. On the other hand there is only `E` in error channel so by types it looks OK, to collect only one (first to complete promise) failure.
Other problem for me is when effects are `uninterruptible` - should we then also await completion of other effects even if we discard their results? Sequential analog would be foreach not short-circuiting on failure.

That said, I'm going to implement what I'm told to implement - can you tell me what to do here?
If I have to merge all failures,  I'll try `Array[Exit[E, A]]` initially filled with `Failure(Cause.interrupt)` and also ignored during merge - WDYT about this?
In case of successes I can collect to List.

Reply:
> Other problem for me is when effects are uninterruptible - should we then also await completion of other effects even if we discard their results? Sequential analog would be foreach not short-circuiting on failure.

We already always wait for completion – in case of success we must wait for all results, in case of failure `.onError(_ => Fiber.interruptAll)` will wait for all fibers to interrupt. This is consistent with current behavior of `foreachPar`, and with `race`, `zipWithPar` and other concurrent ops in ZIO, so it makes sense to me to do this here too!

> Sequential analog would be foreach not short-circuiting on failure.

This <strikethrough>is</strikethrough> was one of the shortcomings of `interruption inheritance` – `foreachPar(_).uninterruptible` behaves strange - https://github.com/zio/zio/issues/945#issuecomment-498068296 – this PR fixes that by wrapping workers with `interruptible`, but obviously you can reintroduce it with another `uninterruptible` inside – but in that case it'll behave same as all `race*` variants, `zipWithPar` and others so I think it's completely correct.

> Now I'm wondering if previous behavior was intentional or just accidental because of implementation with zipWithPar.

It's intentional, I'm sure. Being lossless about errors has always been one of the design principles

> On the other hand there is only E in error channel so by types it looks OK, to collect only one (first to complete promise) failure.

Well, `Cause` like `List` can always contain multiple E's without having to change the type so that's ok

Reply:
@LGLO

Reply:
@LGLO 
> I'll try Array[Exit[E, A]] initially filled with Failure(Cause.interrupt) and also ignored during merge - WDYT about this? In case of successes I can collect to List.

I feel like pre-filling the Array is not necessary, since you either get a single element on success or wait for it to fill up completely on failure anyway.

Reply:
Thank you hundredfold for this detailed response!
I'll do what you said.
I still have a problem with StacktracesSpec, but I'll try to understand why and eventually ask for help.

Reply:
@LGLO Re: StacktracesSpec, I'd advise you turn `val debug = true` in the test, inspect the traces that fail and if they look fine on manual inspection, just change the predicate in failed tests to accept them. – the traces are expected to change with new implementation, the only thing you can do here is manually check that they're useful and then change the test expectation